README.md
Fifty-two Activity 
===============

FiftyTwo is a set of card games, such as Go Fish, Crazy Eights, Hearts, Solitaire, and maybe eventually games like Eucher or Spades.

How to use?
===============
Fifty-two is not a part of Sugar Desktop but can be added. Refer to the following links-

* [How to Get Sugar on sugarlabs.org](https://sugarlabs.org/), 
* [How to use Sugar](https://help.sugarlabs.org/)

<img src="screenshots/2.png" width="75%" title="Fifty-two Activity">

How to upgrade?
===============
On Sugar Desktop systems;

* [Use My Settings,](https://help.sugarlabs.org/my_settings.html) [Software Update](https://help.sugarlabs.org/my_settings.html#software-update) 
* [Use Browse to open ](https://activities.sugarlabs.org/)activities.sugarlabs.org Search for fifty-two, then download


NEWS
All entries are in DD/MM/YY format

v4
7/08/20 - Release
 - Add README.md, Refresh POT file, update activity.info

7/08/20 - Fixes
 - Stop the activity with Gtk stop button, Remove pygame.init() from run(), Fix race condition

9/01/20 - Port to Python3

6/08/18 - Add license and update activity.info

22/02/18 - Port Activity
 - Port activity to GTK3, Sugar Toolkit 3 and Sugargamev1.2

v3
23/06/08

FiftyTwo now has many features, including: A main menu, photo decks, support for the newer joyride builds, and a fully-functional GUI. This is version 0.2.

05/03/08 - Secondary commit
Crazy Eights is pretty much complete for non-mesh play. The FiftyTwo library leaves a lot to be desired for, though. It's pretty messy and needs optimization and doc comments. The smooth movement uses dirty rectangles now, rather than simply clearing the entire screen, making it several times faster.

The AI play has been improved somewhat, but the suit-selection when the AI plays an eight needs to be fixed.

Overall, Crazy Eights isn't doing too badly.

06/02/08 - Initial release
Crazy Eights is now playable in sugar form, though the AI still leaves something to be desired for.

Currently, when you launch FiftyTwo, it goes directly into a game of Crazy Eights, 2 player, with you on the bottom and an AI player on the top. I'm working on a "setup game" menu that should allow you to set up a game of, say, AI versus AI, or player on the top and AI on the bottom, etc.

Also, smooth movement is, shall we say, a tad slow?




gui.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

############################################################################
#    Copyright (C) 2007 by kawk                                            #
#    kawk@theprogrammingsite.com                                           #
#                                                                          #
#    This file is a generic GUI library for the FiftyTwo card game set.    #
#    FiftyTwo is a set of card games for the OLPC XO laptop.               #
#                                                                          #
#    This program is free software; you can redistribute it and/or modify  #
#    it under the terms of the GNU General Public License as published by  #
#    the Free Software Foundation; either version 2 of the License, or     #
#    (at your option) any later version.                                   #
#                                                                          #
#    This program is distributed in the hope that it will be useful,       #
#    but WITHOUT ANY WARRANTY; without even the implied warranty of        #
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         #
#    GNU General Public License for more details.                          #
#                                                                          #
#    You should have received a copy of the GNU General Public License     #
#    along with this program; if not, write to the Free Software           #
#    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 #
#    USA                                                                   #
############################################################################

import pygame
from pygame.locals import *

from gettext import gettext as _

import fiftytwo

class widget(pygame.sprite.Sprite):
    def __init__(self, position, size):
        pygame.sprite.Sprite.__init__(self)
        self.visible = 0
        self.rect = Rect(((0, 0), (0, 0)))
        self.rect.move_ip(position)
        self.rect.w = size[0]
        self.rect.h = size[1]
        self.special = 0

    def detect_click(self, mouseposition):
        if mouseposition[0] > self.rect.left and mouseposition[0] < self.rect.right:
            if mouseposition[1] > self.rect.top and mouseposition[1] < self.rect.bottom:
                return True

        return False

class container(pygame.sprite.OrderedUpdates):
    def __init__(self, widgets, position, size):
        pygame.sprite.OrderedUpdates.__init__(self, widgets)
        self.rect = Rect((position, size))

    def draw(self, surface, mouseposition):
        for s in self.sprites():
            if s.special and mouseposition[0] > s.rect.left and mouseposition[0] < s.rect.right \
                and mouseposition[1] > s.rect.top and mouseposition[1] < s.rect.bottom:
                    if mouseposition[2] is not 0:
                        surface.blit(s.clickimage, s.rect)
                    else:
                        surface.blit(s.hoverimage, s.rect)
            else:
                surface.blit(s.image, s.rect)

class label(widget):
    def __init__(self, position, size, title, color, textsize=38):
        widget.__init__(self, position, size)
        font = pygame.font.Font(None, textsize)
        self.image = font.render(title, 1, color)
        self.image = self.image.convert_alpha()
        self.rect = self.image.get_rect()
        self.rect.x = position[0]
        self.rect.y = position[1]


class image(widget):
    def __init__(self, position, size, file, trans=None):
        widget.__init__(self, position, size)
        self.image, self.rect = fiftytwo.load_image(file, trans)
        self.rect.x = position[0]
        self.rect.y = position[1]
        self.rect.w = size[0]
        self.rect.h = size[1]

class button(widget):
    def __init__(self, position, size, title):
        widget.__init__(self, position, size)
        self.special = 1
        self.label = label(position, size, title, (255, 255, 255))
        self.image = pygame.Surface(size)
        self.hoverimage = pygame.Surface(size)
        self.clickimage = pygame.Surface(size)
        leftrect = self.image.get_rect()
        leftrect.x = 0
        leftrect.y = 0
        leftrect.w = 32
        leftrect.h = size[1]
        pygame.draw.ellipse(self.image, (128, 128, 128), leftrect)
        pygame.draw.ellipse(self.hoverimage, (176, 176, 176), leftrect)
        pygame.draw.ellipse(self.clickimage, (208, 208, 208), leftrect)
        leftrect.x = size[0]-32
        pygame.draw.ellipse(self.image, (128, 128, 128), leftrect)
        pygame.draw.ellipse(self.hoverimage, (176, 176, 176), leftrect)
        pygame.draw.ellipse(self.clickimage, (208, 208, 208), leftrect)
        self.image.fill((128, 128, 128), ((16, 0), (size[0]-32, size[1])))
        self.hoverimage.fill((176, 176, 176), ((16, 0), (size[0]-32, size[1])))
        self.clickimage.fill((208, 208, 208), ((16, 0), (size[0]-32, size[1])))
        self.image.blit(self.label.image, (10, 4))
        self.hoverimage.blit(self.label.image, (10, 4))
        self.clickimage.blit(self.label.image, (10, 4))
        self.image.set_colorkey((0, 0, 0), pygame.RLEACCEL)
        self.hoverimage.set_colorkey((0, 0, 0), pygame.RLEACCEL)
        self.clickimage.set_colorkey((0, 0, 0), pygame.RLEACCEL)

class frame(container):
    def __init__(self, widgets, position, size):
        container.__init__(self, widgets, position, size)

    def draw(self, surface, mouseposition):
        container.draw(self, surface, mouseposition)
        pygame.draw.rect(surface, (32, 32, 32), self.rect, 5)

class popup(frame):
    def __init__(self, widgets, position, size):
        frame.__init__(self, widgets, position, size)

    def draw(self, surface, mouseposition):
        surface.fill((48, 48, 48), self.rect)
        frame.draw(self, surface, mouseposition)


activity.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

import sys
import pygame
import pygame.camera

import gi
gi.require_version('Gtk', '3.0')
from gi.repository import Gtk
from gi.repository import Gdk

from sugar3.activity import activity
from sugar3.graphics.toolbarbox import ToolbarBox
from sugar3.activity.widgets import ActivityToolbarButton
from sugar3.activity.widgets import StopButton
from sugar3.graphics.objectchooser import ObjectChooser
from sugar3.graphics.style import GRID_CELL_SIZE

import sugargame.canvas

import run

from gettext import gettext as _

class Activity(activity.Activity):

    def __init__(self, handle):
        activity.Activity.__init__(self, handle)

        self.game = run.Game(self)
        self.build_toolbar()
        self.game.canvas = self._pygamecanvas = sugargame.canvas.PygameCanvas(
            self, main=self.game.run, modules=[pygame.display, pygame.font,
            pygame.camera])
        self.set_canvas(self._pygamecanvas)
        Gdk.Screen.get_default().connect('size-changed',
                                         self.__configure_cb)

    def __configure_cb(self, event):
        ''' Screen size has changed '''
        pygame.display.set_mode((Gdk.Screen.width(),
                                 Gdk.Screen.height() - GRID_CELL_SIZE),
                                pygame.RESIZABLE)

    def build_toolbar(self):
        toolbar_box = ToolbarBox()
        self.set_toolbar_box(toolbar_box)
        toolbar_box.show()

        activity_button = ActivityToolbarButton(self)
        toolbar_box.toolbar.insert(activity_button, -1)
        activity_button.show()

        separator = Gtk.SeparatorToolItem()
        separator.props.draw = False
        separator.set_expand(True)
        toolbar_box.toolbar.insert(separator, -1)
        separator.show()

        stop_button = StopButton(self)
        toolbar_box.toolbar.insert(stop_button, -1)
        stop_button.show()
        stop_button.connect('clicked', self._stop_cb)


    def read_file(self, file_path):
        pass

    def write_file(self, file_path):
        pass

    def load_image_from_journal(self):
        chooser = None
        name = None
        jobject = None
        try:
            chooser = ObjectChooser(parent=self, what_filter=None)
        except TypeError:
            chooser = ObjectChooser(
                None, self,
                Gtk.DialogFlags.MODAL | 0)
        if chooser is not None:
            try:
                result = chooser.run()
                if result == Gtk.ResponseType.ACCEPT:
                    jobject = chooser.get_selected_object()
                    if jobject and jobject.file_path:
                        name = jobject.metadata['title']
                        #mime_type = jobject.metadata['mime_type']
                        #_logger.debug('result of choose: %s (%s)' % \(name, str(mime_type)))
            finally:
                chooser.destroy()
                del chooser
        
        return jobject

    def _stop_cb(self, button):
        self.game.running = False


crazyeights.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

############################################################################
#    Copyright (C) 2007 by kawk                                            #
#    kawk@theprogrammingsite.com                                           #
#                                                                          #
#    This file is a Crazy Eights card game for the FiftyTwo card game set. #
#    FiftyTwo is a set of card games for the OLPC XO laptop.               #
#                                                                          #
#    This program is free software; you can redistribute it and/or modify  #
#    it under the terms of the GNU General Public License as published by  #
#    the Free Software Foundation; either version 2 of the License, or     #
#    (at your option) any later version.                                   #
#                                                                          #
#    This program is distributed in the hope that it will be useful,       #
#    but WITHOUT ANY WARRANTY; without even the implied warranty of        #
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         #
#    GNU General Public License for more details.                          #
#                                                                          #
#    You should have received a copy of the GNU General Public License     #
#    along with this program; if not, write to the                         #
#    Free Software Foundation, Inc.,                                       #
#    51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA               #
############################################################################

import gi
gi.require_version('Gtk', '3.0')
from gi.repository import Gtk

import fiftytwo
from fiftytwo import card_width, card_height, UP, DOWN
import math, random
import pygame
from pygame.locals import *

from gettext import gettext as _

import run

LOCALPLAYER = 0
AIPLAYER = 1
REMOTEPLAYER = 2

BGCOL = (0, 32, 64)

class player(fiftytwo.cardgroup):
    """
    Basic player class -- not much to it. Inherits from fiftytwo.cardgroup
    """
    def __init__(self, position):
        fiftytwo.cardgroup.__init__(self, (), position[0], position[1])
        self.name = _("Unnamed")
        self.number = 0

    def next_turn(self, topcardid):
        pass

    def render_name(self):
        font = pygame.font.Font(None, 28)
        self.id = font.render(self.name, 1, (255, 255, 0))

class localplayer(player):
    def __init__(self, position):
        player.__init__(self, position)
        self.ai = 0
        self.type = LOCALPLAYER

    def next_turn(self, topcardid):
        return 0

class remoteplayer(player):
    def __init__(self, position):
        player.__init__(self, position)
        self.ai = 0
        self.type = REMOTEPLAYER

    def next_turn(self, topcardid):
        return 0

class aiplayer(player):
    def __init__(self, position):
        player.__init__(self, position)
        self.ai = 1
        self.type = AIPLAYER

    def next_turn(self, topcardid, currentsuit):
        list = []
        for tc in self.sprites():
            if check_valid(topcardid, tc.cardid, currentsuit):
                list.insert(0, tc)

        highvalue = 0
        highcard = None
        for tc in list:
            if tc.cardid == 'JS':
                highvalue = 100
                highcard = tc
            elif tc.cardid[0] == '8':
                if highvalue <= 1:
                    highvalue = 1
                    highcard = tc
            elif tc.cardid[0] == '2':
                if highvalue <= 75 and tc.cardid[1] != topcardid[1]:
                    highvalue = 50
                    highcard = tc
                elif highvalue <= 75 and tc.cardid[1] == topcardid[1]:
                    highvalue = 76
                    highcard = tc
            else:
                highcard = tc

        return highcard

    def choose_new_suit(self):
        """
        Chooses a new suit based on the player's current hand.

        BUGS: If the player has an equal number of different suits, the
        first suit in the list "1, 2, 3, 4" will be chosen.
        """
        list = [['1', 0], ['2', 0], ['3', 0], ['4', 0]]

        for sprite in self.sprites():
            if sprite.cardid[0] == '8':
                continue

            if sprite.cardid[1] == '1':
                list[0][1] += 1
            elif sprite.cardid[1] == '2':
                list[1][1] += 1
            elif sprite.cardid[1] == '3':
                list[2][1] += 1
            elif sprite.cardid[1] == '4':
                list[3][1] += 1

        highest = ['!', 0]
        for x in list:
            if x[1] > highest[1]:
                highest[1] = x[1]
                highest[0] = x[0]

        return highest[0]


def check_valid(topcard, tocheck, currentsuit):
    if tocheck[0] == '8':
        return 1
    elif topcard[0] == '8':
        if tocheck[1] == currentsuit:
            return 1
        else:
            return 0
    elif tocheck[1] == topcard[1]:
        return 1
    elif tocheck[0] == topcard[0]:
        return 1
    else:
        return 0

def end_turn(players, whosturn):
    whosturn += 1
    whosturn = int(math.fmod(whosturn, len(players)))
    return whosturn

def prev_turn(players, whosturn):
    whosturn -= 1
    whosturn = int(math.fmod(whosturn, len(players)))
    return whosturn


def choose_suit(player, choosesuitimage, discard):
    screen = pygame.display.get_surface()
    font = pygame.font.Font(None, 28)
    text = font.render(_("Choose a suit"), 1, (0, 255, 255))

    screen.fill((128, 64, 64), ((50, 50), (200, 200)))

    fpslimiter = pygame.time.Clock()

    while 1:
        fpslimiter.tick(20)

        while Gtk.events_pending():
            Gtk.main_iteration()

        for event in pygame.event.get():
            if event.type == QUIT:
                return
            elif event.type == MOUSEBUTTONDOWN:
                if event.pos[0] < 150 and event.pos[0] > 100 and event.pos[1] > 100 and event.pos[1] < 150:
                    screen.fill(BGCOL, ((50, 50), (250, 250)))
                    return '1'

                elif event.pos[0] < 200 and event.pos[0] > 150 and event.pos[1] > 100 and event.pos[1] < 150:
                    screen.fill(BGCOL, ((50, 50), (250, 250)))
                    return '2'

                elif event.pos[0] < 150 and event.pos[0] > 100 and event.pos[1] > 150 and event.pos[1] < 200:
                    screen.fill(BGCOL, ((50, 50), (250, 250)))
                    return '3'
                elif event.pos[0] < 200 and event.pos[0] > 150 and event.pos[1] > 150 and event.pos[1] < 200:
                    screen.fill(BGCOL, ((50, 50), (250, 250)))
                    return '4'

        screen.blit(choosesuitimage, (100, 100))
        screen.blit(text, (100, 50))
        clear_prev(discard)
        clear_prev(player)
        discard.next_frame(screen)
        player.next_frame(screen)

        pygame.display.flip()

    screen.fill(BGCOL, ((50, 50), (250, 250)))

def clear_prev(group):
    screen = pygame.display.get_surface()
    for s in group.sprites():
        screen.fill(BGCOL, s.get_prevrect())

def special_cards(deck, discard, players, whosturn, cardprocessed, choosesuitimage, currentsuit):
    if discard.sprites()[-1].cardid[0] == '2' and cardprocessed is 0:
        fiftytwo.draw_card(deck, players[whosturn])
        if players[whosturn].ai == 0: players[whosturn].sprites()[-1].set_face(UP)
        fiftytwo.draw_card(deck, players[whosturn])
        if players[whosturn].ai == 0: players[whosturn].sprites()[-1].set_face(UP)
        cardprocessed = 1

    elif discard.sprites()[-1].cardid[0] == '8' and cardprocessed is 0:
        whosturn = prev_turn(players, whosturn)

        if players[whosturn].ai is not 1:
            currentsuit = choose_suit(players[whosturn], choosesuitimage, discard)
        else:
            currentsuit = players[whosturn].choose_new_suit()
        cardprocessed = 1

        whosturn = end_turn(players, whosturn)

    elif discard.sprites()[-1].cardid == 'J4' and cardprocessed is 0:
        whosturn = end_turn(players, whosturn)

        cardprocessed = 1

    elif cardprocessed is 1:
        pass

    else:
        cardprocessed = 1

    return cardprocessed, whosturn, currentsuit

def main(playertypes, screensize):
    running = True
    sw = screensize[0]
    sh = screensize[1]

    localplayers = 0

    # Reduce load
    pygame.event.set_blocked(pygame.MOUSEMOTION)

    #positions = (((5, -card_height()/2), (sw-5, card_height()-(card_height()/2))), ((5, sh-card_height()/2), (sw-5, sh+card_height()/2)))
    positions = (((int(card_width()*1.5), -card_height()/2), (sw-int(card_width()*1.5), card_height()-(card_height()/2))),
        ((sw-int(card_width()*0.5), 5), (sw+int(card_width()*0.5), sh-5)),
        ((int(card_width()*1.5), sh-card_height()/2), (sw-int(card_width()*1.5), sh+card_height()/2)),
        ((-int(card_width()*0.5), 5), (int(card_width()*0.5), sh-5)))


    idpositions = ((sw/2, card_height()*0.75), (sw-card_width()*1.5+15, int(sh*(2/3.0))),
        (sw/2, sh-card_height()*0.75), (card_width()*0.75, int(sh*(2/3.0))))

    screen = pygame.display.get_surface()

    font = pygame.font.Font(None, 48)

    loading = font.render(_("Loading, please wait!"), 1, (255, 255, 255))

    screen.blit(loading, (450, 350))

    pygame.display.flip()

    suit1, temprect = fiftytwo.load_image('suit1.' + run.deck.deck + '.png', -1)
    suit2, temprect = fiftytwo.load_image('suit2.' + run.deck.deck + '.png', -1)
    suit3, temprect = fiftytwo.load_image('suit3.' + run.deck.deck + '.png', -1)
    suit4, temprect = fiftytwo.load_image('suit4.' + run.deck.deck + '.png', -1)

    deck = fiftytwo.cardgroup(fiftytwo.shuffledeck(), (sw / 6, (sh - card_height()) / 2),
        (sw / 6 + sw / 4, (sh + card_height()) / 2))
    discard = fiftytwo.cardgroup((), (sw - int(card_width()*3), (sh - card_height()) / 2),
        (sw - int(card_width()*1.5), (sh + card_height()) / 2))

    players = []

    num = -1

    for c in playertypes:
        num += 1

        if len(playertypes) == 2 and num == 1:
            num = 2

        if c[0] == 'l':
            players += [localplayer(positions[num])]
            players[-1].name = c[1:len(c)]
            players[-1].number = num
            players[-1].render_name()
            localplayers += 1

        elif c[0] == 'a':
            players += [aiplayer(positions[num])]
            players[-1].name = c[1:len(c)]
            players[-1].number = num
            players[-1].render_name()

    fpslimiter = pygame.time.Clock()

    #choosesuitimage, unused = fiftytwo.load_image('suits.' + run.deck.deck + '.png', -1)
    choosesuitimage = pygame.Surface((suit1.get_width()*2, suit1.get_height()*2))
    choosesuitimage.fill((255, 255, 255))
    choosesuitimage.set_colorkey((255, 255, 255), pygame.RLEACCEL)
    choosesuitimage.blit(suit1, (0, 0))
    choosesuitimage.blit(suit2, (suit1.get_width(), 0))
    choosesuitimage.blit(suit3, (0, suit1.get_height()))
    choosesuitimage.blit(suit4, (suit1.get_width(), suit1.get_height()))

    fiftytwo.draw_card(deck, discard)
    cardprocessed = 0
    currentsuit = discard.sprites()[0].cardid[1]
    discard.sprites()[0].set_face(UP)

    for p in players:
        for x in range(0, 8):
            fiftytwo.draw_card(deck, p)
            p.sprites()[-1].set_face(UP)


    whosturn = random.randint(0, len(playertypes)-1)

    for p in players:
        if p.type == LOCALPLAYER:
            p.set_all_faces(UP)
        elif p.type == AIPLAYER:
            p.set_all_faces(DOWN)

    screen.fill(BGCOL)

    while running:
        fpslimiter.tick(20)

        # Check for win
        for x in players:
            if len(x.sprites()) == 0:
                screen.fill(BGCOL)
                gameover = font.render(_("Game over,"), 1, (0, 255, 255))
                m = _('%s won the game!')
                message = font.render(m % x.name, 1, (255, 255, 0))

                for p in players:
                    p.set_all_faces(UP)
                    p.next_frame(screen)
                    pygame.display.flip()

                for x in range(100):
                    fpslimiter.tick(15)

                    clear_prev(deck)
                    clear_prev(discard)
                    for p in players:
                        clear_prev(p)

                    deck.next_frame(screen)
                    discard.next_frame(screen)

                    for p in players:
                        p.next_frame(screen)

                    screen.fill((32, 64, 64), (((sw/2)-(600/2), (sh/2)-(100/2)), (600, 100)) )
                    screen.blit(gameover, ((sw/2)-((600/2)-((600-gameover.get_rect().w)/2)), (sh/2)-50))
                    screen.blit(message, ((sw/2)-((600/2)-((600-message.get_rect().w)/2)), (sh/2)))

                    pygame.display.flip()

                    for event in pygame.event.get():
                        if event.type == KEYDOWN:
                            if event.key == K_ESCAPE:
                                return

                return

        if len(deck.sprites()) is 0:
            while len(discard.sprites()) > 1:
                sprite = discard.sprites()[0]
                sprite.set_face(DOWN)
                sprite.move_to_group(deck)

            deck.shuffle()

        while Gtk.events_pending():
            Gtk.main_iteration()

        if not running:
            break

        for event in pygame.event.get():

            if event.type == KEYDOWN:
                if event.key == K_ESCAPE:
                    return
                elif event.key == K_SPACE:
                    for s in discard.sprites():
                        s.set_pos(s.gotox, s.gotoy)

                    for s in deck.sprites():
                        s.set_pos(s.gotox, s.gotoy)

                    for p in players:
                        for s in p.sprites():
                            s.set_pos(s.gotox, s.gotoy)

                    screen.fill(BGCOL)

            elif event.type == MOUSEBUTTONDOWN:
                if players[whosturn].type != LOCALPLAYER:
                    continue
                r = players[whosturn].detect_click(event.pos)
                if r is not None:
                    if check_valid(discard.sprites()[-1].cardid, r.cardid, currentsuit):
                        r.move_to_group(discard)
                        whosturn = end_turn(players, whosturn)

                        currentsuit = discard.sprites()[-1].cardid[1]

                        cardprocessed = 0
                else:
                    r = deck.detect_click(event.pos)
                    if r is not None and players[whosturn].type == LOCALPLAYER:
                        fiftytwo.draw_card(deck, players[whosturn])
                        players[whosturn].sprites()[-1].set_face(UP)
                        whosturn = end_turn(players, whosturn)

            elif event.type == QUIT:
                running = False
                return running

            """elif event.type == olpcgames.CONNECT:
                print "Connect event."
                pass

            elif event.type == olpcgames.PARTICIPANT_ADD:
                print "Add event."
                pass

            elif event.type == olpcgames.PARTICIPANT_REMOVE:
                print "Remove event."
                pass"""

        if players[whosturn].ai is 1 and cardprocessed is 1:
            cready = 1

            for p in players:
                if p.check_ready() == 0:
                    cready = 0

            if discard.check_ready() == 0:
                cready = 0

            if deck.check_ready() == 0:
                cready = 0

            if cready == 1:
                r = players[whosturn].next_turn(discard.sprites()[-1].cardid, currentsuit)
                if r is None:
                    fiftytwo.draw_card(deck, players[whosturn])
                    whosturn = end_turn(players, whosturn)
                else:
                    r.move_to_group(discard)
                    r.set_face(UP)
                    cardprocessed = 0
                    whosturn = end_turn(players, whosturn)
                    currentsuit = r.cardid[1]

        # Gameplay stuff

        cardprocessed, whosturn, currentsuit = special_cards(deck, discard, players, whosturn, cardprocessed, choosesuitimage, currentsuit)

        clear_prev(deck)
        clear_prev(discard)
        for p in players:
            #p.sort()
            clear_prev(p)

        if currentsuit == '1':
            screen.fill(BGCOL, (((sw/2)-(suit1.get_width()/2), (sh/2)-(suit1.get_height()/2)), (suit1.get_width(), suit1.get_height())))
            screen.blit(suit1, ((sw/2)-(suit1.get_width()/2), (sh/2)-(suit1.get_height()/2)))
        elif currentsuit == '2':
            screen.fill(BGCOL, (((sw/2)-(suit2.get_width()/2), (sh/2)-(suit2.get_height()/2)), (suit2.get_width(), suit2.get_height())))
            screen.blit(suit2, ((sw/2)-(suit2.get_width()/2), (sh/2)-(suit2.get_height()/2)))
        elif currentsuit == '3':
            screen.fill(BGCOL, (((sw/2)-(suit3.get_width()/2), (sh/2)-(suit3.get_height()/2)), (suit3.get_width(), suit3.get_height())))
            screen.blit(suit3, ((sw/2)-(suit3.get_width()/2), (sh/2)-(suit3.get_height()/2)))
        elif currentsuit == '4':
            screen.fill(BGCOL, (((sw/2)-(suit4.get_width()/2), (sh/2)-(suit4.get_height()/2)), (suit4.get_width(), suit4.get_height())))
            screen.blit(suit4, ((sw/2)-(suit4.get_width()/2), (sh/2)-(suit4.get_height()/2)))

        deck.next_frame(screen)
        for p in players:
            p.next_frame(screen)
            screen.blit(p.id, idpositions[p.number])

        discard.next_frame(screen)

        pygame.display.flip()
